_posts/2016-08-26-babili.md

```diff
@@ -0,0 +1,129 @@
+
+```js
+// ES6 code -> Babili -> Minified ES6 Code
+class a{constructor(b){this.program=b}}new a;
+```
+
+Also it's important to note that this isn't specific to ES6. Because Babel updates as ECMAScript updates (with [ES2015, ES2016, and now ES2017](http://babeljs.io/docs/plugins/#official-presets)) and follows the proposal process for experimental features (with our [stage-x presets](http://babeljs.io/docs/plugins/#stage-x-experimental-presets)), the minifier should be able to output whatever version of ECMAScript that is supported.
+
+## Usage
+
+If you already use Babel, you can just add the [babili](https://github.com/babel/babili#babel-preset) preset (babel-preset-babili) to your config.
+
+You probably want to enable this only in production, so use the [env option](http://babeljs.io/docs/usage/babelrc/#env-option) which uses either `process.env.BABEL_ENV` or `process.env.NODE_ENV`
+
+```bash
+$ npm install babel-preset-babili --save-dev
+```
+
+```js
+// previous .babelrc
+{ "presets": ["es2015"] }
+// .babelrc
+{
+  "presets": ["es2015"],
+  "env": {
+    "production": {
+      "presets": ["babili"]
+    }
+  }
+}
+```
+
+If you don't use Babel, you can use our standalone cli tool [`babili`](https://github.com/babel/babili#cli). (Currently it's just a wrapper for babel-cli + the preset)
+
+```bash
+$ npm install babili --save-dev
+```
+
+```bash
+# equivalent to 
+# babel src -d lib --presets=babili
+$ babili src -d lib
+```
+
+## Pros/Cons
+
+Uglify Pros
+- no change to existing tooling if you are already minifying
+- battle-tested/production ready (been around for years with many users so there's less bugs/issues).
+- fast, outputs small code already
+
+Uglify Cons
+- Custom parser/tooling, so difficult to output/minify ES6 and more.
+- Not modular, no way to create own plugins/minification strategies
+
+Babili Pros:
+- ES6+ aware (nothing special needs to be done because we can use the babylon parser) and Babel will update as standards/browsers update.
+- Uses the existing babel toolchain, can consume as a babel preset or standalone.
+- Potential to work on integrating better with existing bundlers like webpack.
+- Each minfication step can be split into into it's own plugins and plenty of options for customization. Makes it easier to both contribute and to find/submit issues for specific problems. Also people can be free to experiment with their own plugins and get them accepted into core.
+  - For example: [this](https://github.com/babel/babili/tree/master/packages/babel-plugin-transform-minify-booleans) just turns `true` into `!0` which is straightforward to write.
+- Should be an easy transition if people are familiar with transpiling with Babel already.
+
+Babili Cons:
+- Just released (early), so not many users yet. Early adopters will have to deal with tool that isn't as battle-tested as Uglify at first.
+- At the moment, might not be faster or produce smaller gziped sizes than Uglify or other tools.
```

Comment:
I think we should be straightforward about our shortcomings and say that performance/size are inferior at the moment (not just might; our tests show that it's worse on various files) but that we're focusing on improving that.